ibis/expr/analysis.py

```diff
@@ -222,26 +222,9 @@ def substitute_unbound(node):
     def fn(node):
         if isinstance(node, ops.DatabaseTable):
             return ops.UnboundTable(name=node.name, schema=node.schema)
-        elif isinstance(node, ops.TableColumn):
-            # For table column references, in the event that we're on top of a
-            # projection, we need to check whether the ref comes from the base
-            # table schema or is a derived field. If we've projected out of
-            # something other than a physical table, then lifting should not
-            # occur
-            table = node.table
-
-            if isinstance(table, ops.Selection):
-                for val in table.selections:
-                    if isinstance(val, ops.PhysicalTable) and node.name in val.schema:
-                        return ops.TableColumn(val, node.name)
-        elif isinstance(node, ops.Join):
-            return node.__class__(
-                substitute_unbound(node.left),
```

Comment:
The substitution happens in depth-first order so these nodes should have database tables already replaced.

Reply:
For whatever reason, this doesn't descend into join predicates and also leaves some `TableColumn`s bound to the previous parent table.

Reply:
Do we have tests for that case?

Reply:
Probably because the join predicates don't have proper rules set: `Join.predicates = rlz.optional(lambda x, this: x, default=())`, so it's a tuple instead of a nodelist.

Reply:
Ahh, yeah, it was definitely tuple, hence the `map` call in my original version

Reply:
Sadly we still have missing rule definitions scattered across the codebase.